src/serde/de_br.rs

```diff
@@ -126,11 +271,93 @@ mod tests {
         let node = node_from_bytes_backrefs(&mut allocator, &buf).unwrap();
+        let old_node = node_from_bytes_backrefs_old(&mut allocator, &buf).unwrap();
         let mut oc = ObjectCache::new(treehash);
         let calculated_hash = oc.get_or_calculate(&allocator, &node, None).unwrap();
         let ch: &[u8] = calculated_hash;
         let expected_hash: Vec<u8> = Vec::from_hex(expected_hash_as_hex).unwrap();
         assert_eq!(expected_hash, ch);
+        let calculated_hash = oc.get_or_calculate(&allocator, &old_node, None).unwrap();
+        let ch: &[u8] = calculated_hash;
+        assert_eq!(expected_hash, ch);
+    }
+
+    #[cfg(feature = "counters")]
+    #[test]
+    fn test_counters() {
+        use crate::allocator::Allocator;
+
+        let mut a = Allocator::new();
+        let cp = a.checkpoint();
+        a.add_ghost_pair(1).unwrap();
+        assert_eq!(a.pair_count(), 1);
+
+        a.restore_checkpoint(&cp);
+        let buf = Vec::from_hex("0a").unwrap();
+        let _node = node_from_bytes_backrefs(&mut a, &buf).unwrap();
+        let pair_count = a.pair_count();
+        let pair_count_no_ghosts = a.pair_count_no_ghosts();
+        a.restore_checkpoint(&cp);
+        let _old_node = node_from_bytes_backrefs_old(&mut a, &buf).unwrap();
+        let old_pair_count = a.pair_count();
+        let old_pair_count_no_ghosts = a.pair_count_no_ghosts();
+        assert_eq!(pair_count_no_ghosts, 0);
+        assert_eq!(old_pair_count_no_ghosts, 1);
+        assert_eq!(pair_count, old_pair_count);
+    }
+
+    #[test]
+    fn test_traverse_path_with_vec() {
+        use crate::allocator::Allocator;
+
+        let mut a = Allocator::new();
+        let nul = a.nil();
+        let n1 = a.new_atom(&[0, 1, 2]).unwrap();
+        let n2 = a.new_atom(&[4, 5, 6]).unwrap();
+
+        let mut list: Vec<(NodePtr, Option<NodePtr>)> = vec![(n1, None)];
+
+        assert_eq!(traverse_path_with_vec(&mut a, &[], &mut list).unwrap(), nul);
+
+        // cost for leading zeros
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0], &mut list).unwrap(),
+            nul
+        );
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0, 0], &mut list).unwrap(),
+            nul
+        );
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0, 0, 0], &mut list).unwrap(),
+            nul
+        );
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0, 0, 0, 0], &mut list).unwrap(),
+            nul
+        );
+
+        let mut list: Vec<(NodePtr, Option<NodePtr>)> = vec![(n1, None), (n2, None)];
+
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0b10], &mut list).unwrap(),
+            n2
+        );
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0b101], &mut list).unwrap(),
+            n1
+        );
+        assert_eq!(
+            traverse_path_with_vec(&mut a, &[0b111], &mut list).unwrap(),
```

Comment:
I think these cases would be interesting as well:
```
traverse_path_with_vec(&mut a, &[0b1011], &mut list)
traverse_path_with_vec(&mut a, &[0b1111], &mut list)
```

The should both fail as they try to traverse NIL

Reply:
`traverse_path_with_vec(&mut a, &[0b1011], &mut list)` was already tested on line 356, but I've now added `[0b1111]` next to it